src/sentry/issues/auto_source_code_config/task.py

```diff
@@ -180,25 +181,38 @@ def set_project_codemappings(
         )
 
         if not repository:
-            repository = Repository.objects.create(
-                name=code_mapping.repo.name,
-                organization_id=organization_id,
-                integration_id=organization_integration.integration_id,
+            if not dry_run:
+                repository = Repository.objects.create(
+                    name=code_mapping.repo.name,
+                    organization_id=organization_id,
+                    integration_id=organization_integration.integration_id,
+                )
+            metrics.incr(
+                key=f"{METRIC_PREFIX}.repository.created",
+                tags={"platform": platform, "dry_run": dry_run},
+                sample_rate=1.0,
             )
 
-        _, created = RepositoryProjectPathConfig.objects.get_or_create(
-            project=project,
-            stack_root=code_mapping.stacktrace_root,
-            defaults={
-                "repository": repository,
-                "organization_integration_id": organization_integration.id,
-                "integration_id": organization_integration.integration_id,
-                "organization_id": organization_integration.organization_id,
-                "source_root": code_mapping.source_path,
-                "default_branch": code_mapping.repo.branch,
-                "automatically_generated": True,
-            },
-        )
-        if created:
-            # Since it is a low volume event, we can sample at 100%
-            metrics.incr(key="code_mappings.created", tags={"platform": platform}, sample_rate=1.0)
+        # The project and stack_root are unique together
+        configs = RepositoryProjectPathConfig.objects.filter(
+            project=project, stack_root=code_mapping.stacktrace_root
+        ).all()
+        if not configs:
+            if not dry_run and repository is not None:
+                RepositoryProjectPathConfig.objects.create(
+                    repository=repository,
+                    project=project,
+                    organization_integration_id=organization_integration.id,
+                    organization_id=organization_integration.organization_id,
+                    integration_id=organization_integration.integration_id,
+                    stack_root=code_mapping.stacktrace_root,
+                    source_root=code_mapping.source_path,
+                    default_branch=code_mapping.repo.branch,
+                    automatically_generated=True,
+                )
+
+            metrics.incr(
+                key=f"{METRIC_PREFIX}.code_mapping.created",
+                tags={"platform": platform, "dry_run": dry_run},
+                sample_rate=1.0,
+            )
```

Comment:
Shouldn't this be under the `if`? I get that you're not trying to gate it by `dry_run`, but shouldn't it obey the `if repository is not None` check on creation?

Reply:
UPDATE: Okay, I know you changed the logic around, but I'm still not sure I understand it. Shouldn't the `code_mapping.created` only fire when either:

- we're not in dry-run, we call `get_or_create`, and `created` comes back true, or
- we are in dry-run, so we don't call `get_or_create`, but if we _had_ called it, it would have come back `created = True`?

The first one works with the code above, but for the second one, don't we have to make sure the record doesn't yet exist before saying that `created` would have been true?